drivers/ieee802154/ieee802154_rf2xx.c

```diff
@@ -361,16 +361,90 @@ static int rf2xx_set_channel(struct device *dev, u16_t channel)
 	return 0;
 }
 
+/**
+ *  RF output power write
+ *              AT86RF233
+ *
+ *              0 = 4 dBm
+ *              1 = 3.7 dBm
+ *              2 = 3.4 dBm
+ *              3 = 3 dBm
+ *              4 = 2.5 dBm
+ *              5 = 2 dBm
+ *              6 = 1 dBm
+ *              7 = 0 dBm
+ *              8 = -1 dBm
+ *              9 = -2 dBm
+ *              10 = -3 dBm
+ *              11 = -4 dBm
+ *              12 = -6 dBm
+ *              13 = -8 dBm
+ *              14 = -12 dBm
+ *              15 = -17 dBm
+ *              AT86RF212B
+ *              See datasheet for TX power settings
+ *
+ */
 static int rf2xx_set_txpower(struct device *dev, s16_t dbm)
 {
```

Comment:
@markus-becker-tridonic-com thank you for the initiative,
We need keep in mind that this method must works for all devices 231, 232, 233, 212 and in future 215 that is multi transceiver Sub-Giga and 2.4G

@tbursztyka the set tx power is something that is strange for me. This cannot reflect a true radio with external PA for instance. We set 4 dBm but in pratical can be 20 dBm. I was wondering if make sense an API change here. So, we can send an power index instead a dBm level. That way, we can create a table too. I 'll suggest 16 levels but I'm suspect here rsrs.

With that table, upper layers can read the table and adjust as apropriated from lower to higher power or vice-versa.

What do you think? Make sense?

Reply:
@karsev did you check whether the parameters are the same for the other chips of this series?

Reply:
@markus-becker-tridonic-com the real problem happen when there is an external PA/FEM. We need figure out a way to fill up a table to match with the transceiver + PA. for instance, There are FEM that can save more than 40% power in average compared with some SoC vendors at 8dBm or higher. The SKY66408 is new device that have linear output power controlled by I2C with 0.5dBm steps. How we can handle with this external devices?

I mean, there are new things happen that we need to be prepared.

For instance, we can have a storage space that can be have the region's files with the right configuration. That could be a little overkill but we need to be open and look options.

Reply:
@nandojve true, but that applies for all SoCs. currently there is no general mechanism.

Similar approach has been taken in cc13xx_cc26xx and cc2520.

Reply:
@markus-becker-tridonic-com could you send an update with tables like kw41z. That should address active devices supported by driver (231, 232, 233).

Reply:
there's a slight difference between 233 and 231/232 in the TX output powers. the RF233 has a max of 4 dBm and the table is slightly different. should we introduce an ifdef for that or would an approximation be good enough?

```
#if defined(RF233)

#define RF2XX_OUTPUT_POWER_MAX		4
#define RF2XX_OUTPUT_POWER_MIN		(-17)

/* Lookup table for PHY_TX_PWR register */
static const uint8_t phy_tx_pwr_lt[] = {
	0xf,                     /* -17  dBm: -17 */
	0xe, 0xe, 0xe, 0xe, 0xe, /* -12  dBm: -16, -15, -14, -13, -12 */
	0xd, 0xd, 0xd,           /* -9   dBm: -11, -10, -9 */
	0xc, 0xc,                /* -7   dBm: -8, -7 */
	0xb, 0xb,                /* -5   dBm: -6, -5 */
	0xa,                     /* -4   dBm: -4 */
	0x9,                     /* -3   dBm: -3 */
	0x8,                     /* -2   dBm: -2 */
	0x7,                     /* -1   dBm: -1 */
	0x6,                     /*  0.0 dBm:  0 */
	//0x5,                   /*  0.7 dBm */
	0x4,                     /*  1.3 dBm: 1 */
	//0x3,                   /*  1.8 dBm */
	0x2,                     /*  2.3 dBm: 2 */
	//0x1,                   /*  2.8 dBm */
	0x0                      /*  3 dBm: 3 */
};

#else

#define RF2XX_OUTPUT_POWER_MAX		3
#define RF2XX_OUTPUT_POWER_MIN		(-17)

/* Lookup table for PHY_TX_PWR register */
static const uint8_t phy_tx_pwr_lt[] = {
	0xf,                     /* -17  dBm: -17 */
	0xe, 0xe, 0xe, 0xe, 0xe, /* -12  dBm: -16, -15, -14, -13, -12 */
	0xd, 0xd, 0xd, 0xd       /* -8   dBm: -11, -10, -9, -8 */
	0xc, 0xc,                /* -6   dBm: -7, -6 */
	0xb, 0xb,                /* -4   dBm: -5, -4 */
	0xa,                     /* -3   dBm: -3 */
	0x9,                     /* -2   dBm: -2 */
	0x8,                     /* -1   dBm: -1 */
	0x7,                     /*  0.0 dBm:  0 */
	0x6,                     /*  1   dBm:  1 */
	0x5,                     /*  2   dBm:  2 */
	//0x4,                   /*  2.5 dBm */
	0x3,                     /*  3   dBm:  3 */
	//0x2,                   /*  3.4 dBm */
	//0x1,                   /*  3.7 dBm */
	0x0                      /*  4   dBm: 4 */
};

#endif

static int rf2xx_set_txpower(struct device *dev, int16_t dbm)
{
	if (dbm < RF2XX_OUTPUT_POWER_MIN) {
		LOG_INF("TX-power %d dBm below min of %d dBm, using %d dBm",
			    dbm,
			    RF2XX_OUTPUT_POWER_MIN,
			    RF2XX_OUTPUT_POWER_MAX);
		dbm = RF2XX_OUTPUT_POWER_MIN;
	} else if (dbm > RF2XX_OUTPUT_POWER_MAX) {
		LOG_INF("TX-power %d dBm above max of %d dBm, using %d dBm",
			    dbm,
			    RF2XX_OUTPUT_POWER_MIN,
			    RF2XX_OUTPUT_POWER_MAX);
		dbm = RF2XX_OUTPUT_POWER_MAX;
	}

	rf2xx_iface_reg_write(dev, RF2XX_PHY_TX_PWR_REG,
		phy_tx_pwr_lt[dbm - RF2XX_OUTPUT_POWER_MIN]);

	return 0;
}
```

Reply:
@markus-becker-tridonic-com I was thinking about the issue with PA and real dBm on real devices. Since we don't solve it yet and the difference between devices (without external PA) is minimal, I'm open to keep only RF233 table for now and use it for all devices.

Please, send PR and let's move forward.

Reply:
> @markus-becker-tridonic-com I was thinking about the issue with PA and real dBm on real devices. Since we don't solve it yet and the difference between devices (without external PA) is minimal, I'm open to keep only RF233 table for now and use it for all devices.
> 
> Please, send PR and let's move forward.

done